include/DoocsBackendNumericRegisterAccessor.h

```diff
@@ -0,0 +1,371 @@
+            this->buffer_2D[0][i] = ChimeraTK::numericToUserType<UserType>(dataGet<T>(i + this->elementOffset));
+          }
+        });
+    }
+  }
+
+  /********************************************************************************************************************/
+
+  template<>
+  inline void DoocsBackendNumericRegisterAccessor<ChimeraTK::Void>::doPreWrite(
+      TransferType type, VersionNumber version) {
+    DoocsBackendRegisterAccessor<ChimeraTK::Void>::doPreWrite(type, version);
+
+    src.set(0);
+  }
+
+  /********************************************************************************************************************/
+
+  template<typename UserType>
+  void DoocsBackendNumericRegisterAccessor<UserType>::doPreWrite(TransferType type, VersionNumber version) {
+    DoocsBackendRegisterAccessor<UserType>::doPreWrite(type, version);
+
+    // implement read-modify-write
+    if(this->isPartial) {
+      this->doReadTransferSynchronously();
+      this->src = this->dst;
+      // make sure the array is long enough, as the remote server might have changed the length on its side
+      if(size_t(this->src.length()) < this->nElements + this->elementOffset) {
+        this->src.length(this->nElements + this->elementOffset);
+      }
+    }
+
+    // define lambda used below for optimisation
+    auto copyToTargetPointer = [&](auto* targetPointer) {
+      using TargetType = std::remove_reference_t<decltype(*targetPointer)>;
+      assert(targetPointer);
+
+      targetPointer += this->elementOffset;
+      if constexpr(std::is_same<UserType, TargetType>::value) {
+        // raw and target data type match, do a memcopy
+        memcpy(targetPointer, this->buffer_2D[0].data(), this->nElements * sizeof(TargetType));
+      }
+      else {
+        std::transform(this->buffer_2D[0].begin(), this->buffer_2D[0].end(), targetPointer,
+            [](UserType v) { return ChimeraTK::userTypeToNumeric<TargetType>(v); });
+      }
+    };
+
+    // optimise depending on type
+    switch(this->src.type()) {
+      case DATA_SPECTRUM:
+      case DATA_A_FLOAT: {
+        copyToTargetPointer(this->src.get_float_array());
+        return;
+      }
+      case DATA_A_DOUBLE: {
+        copyToTargetPointer(this->src.get_double_array());
+        return;
+      }
+      case DATA_A_INT: {
+        copyToTargetPointer(this->src.get_int_array());
+        return;
+      }
+      case DATA_A_LONG: {
+        copyToTargetPointer(this->src.get_long_array());
+        return;
+      }
+      case DATA_A_BOOL:
+      case DATA_A_SHORT: {
+        copyToTargetPointer(this->src.get_short_array());
+        return;
+      }
+      default:
+        // inefficient copying via single element access for other data types (including scalars)
+        callForDoocsType(this->src, [&](auto t) {
+          using T = decltype(t);
+          if(this->src.array_length() == 0) {
+            // scalar (IFFF, IIII and string are handled in different accessors)
```

Comment:
I understand that string is not handled here because it is the NumericRegisterAccessor. But you read IFFF in this class. Why is this comment in the length == 0 section? Wouldn't the data length be 4 for IFFF and IIII?

Reply:
There is a separate accessor for IFFF and IIII, but they derive from this accessor and do the writing differently. We could integrate this here as well, but I did not want to go that far at this point.

Reply:
Wait, IFFF does not derive from this accessor. We should throw everything out that is related to IFFF I guess.

Reply:
Ah, I was not aware that they derive from here.